### ISSUE TITLE
API 29 example

### ISSUE THREAD

**Issue by RalfWerner** [RalfWerner/integrated-process#2]
## Install APK 
What happens if _my_ `t*.apk` will installed on a _Q/10_ device and opened (bash) or _is it not possible:
`type bash` is **non-native(~/u)?**_ Four shots to install and use of example check see [README](https://github.com/RalfWerner/integrated-process/edit/master/api29):
![grafik](https://user-images.githubusercontent.com/45426380/69047686-0cd64900-09fc-11ea-9138-0adf48579176.png)
- 1 Startscreen with Termux and X-Server plugin on  my current _P/9_ device
- 2 QuickEdit-App to view sources - for comparison with check.
- 3 Termux-App with _process control_ (pp) start/log of example (t/c) alias/env check
- 4 XClient (check.exe) in X-Server to view sources (here *.xpm) -source in preraration.

The action is simple with Android 9. To do the same in 10/API-29 would have to be u/bin u/lib (`ln -s ~/../usr ~/u`) and ~/check.exe ($s) in **_native_ /system**.
@xeffyr said **Yes** to _Could the APK solution be tested on Andriod-9? **How does it work?**_

Github has a limit that my-t77.apk (63Mb) can not be saved - is also [here](https://www.dropbox.com/sh/e34d0nnwingz2c6/AACVbPuTkvXxFgf-LtYy0Ml9a?dl=0) and created with _apksigner_. There are also the versions 0.73 (smaller but needs package update) and 0.82 to find. An exchange of the 'my-t*.apk` (same signing) has no restriction of the check function to follow.

@fornwall made a (proot less) suggestion in [dev](https://gitter.im/termux/dev?at=5dd19664ca778c1fbf1e7634) - where do I find (and how to use) the
 **'termux-package-binary'** and do you've an _example of a **script?**_

**Comment by RalfWerner** [comment]
Github has a limit that my-t77.apk (63Mb) can not be saved - is also [here](https://www.dropbox.com/sh/e34d0nnwingz2c6/AACVbPuTkvXxFgf-LtYy0Ml9a?dl=0) and created with _apksigner_.

**Comment by xeffyr** [comment]
> What happens if my t*.apk will installed on a Q/10 device and opened (bash) or is it not possible:
type bash is non-native(~/u)?

Android prohibits usage of native code where it can be modified. This restriction is enforced for all applications targeting at Android SDK level 29.

Currently Termux targets on SDK 28 and therefore it is not affeced, but will be once we bump used SDK version.

> To do the same in 10/API-29 would have to be u/bin u/lib (ln -s ~/../usr ~/u) and ~/check.exe ($s) in native /system.

No. It has to be in `/data/app/${random dir name}/lib`.
To place binary into /system you need root.

> @xeffyr said Yes to Could the APK solution be tested on Andriod-9? How does it work?

Assume that binary can be executed only from `/data/app/${random dir name}/lib` and its execution causes `permission denied` error if doing this from `/data/data/${package name}`.

Nothing dependent on specific Android versions. Just native code must be placed at paths where it should be.

The closest implementation to one that will be used for Termux is https://github.com/xeffyr/alpine-term/tree/master/alpine-term-app (qemu/socat binaries in jni lib dir and data in dynamic $PREFIX).

> where do I find (and how to use) the 'termux-package-binary' and do you've an example of a script?

Such script is not exist.
But generally, it will be a wrapper around utilities `zip` and `apksigner` for bundling user's binary inside prebuilt APK.

This approach will be the lowest priority target for Termux as soon as proot can be used.

**Comment by RalfWerner** [comment]
> This directory is extracted from APK during application installation process.

In my example 400Mb  created in `u/bin` and `u/lib` etc. with _virgin start_ and package update (also contains **bash**). In `/system/bin` are 8.5Mb and in `*/lib` 185Mb. At least part of the 400Mb would have to be in the `t*.apk` _**right?**
>A: To place binary into `/system` you need root.
B: But generally, it wants to be a wrapper, for bundling user's binary inside _pre**built**_ APK.
C: This approach wants to be the lowest priority target for term as soon as possible.

A: This does not apply if the APK already contains everything or?
B: What would the _termux-build_ look like to do this (+bash etc)?
C: As I understood last Sunday's discourse, the priority should be higher for it to be sustainable.

**Comment by xeffyr** [comment]
> At least part of the 400Mb would have to be in the t*.apk _right?

Right.

> This does not apply if the APK already contains everything or?

Application **cannot** modify the system partition which is created by manufacturer and is read-only.

> What would the termux-build look like to do this (+bash etc)?

It will just append a file to APK, which in fact just a zip archive. Then it will be signed with `apksigner`.

Steps will look like these:
```
cd $TMPDIR
cp $PREFIX/share/user-bin-wrapper/base.zip termux-bin-wrapper.zip
mkdir -p lib/armeabi-v7a/
cp $HOME/your_binary lib/armeabi-v7a/
zip -r9u termux-bin-wrapper.zip lib
apksigner termux-bin-wrapper.zip $HOME/termux-bin-wrapper.apk
```

You will need to install termux-bin-wrapper.apk and open it by clicking on icon to start service exposed to Termux.

Then
```
termux-bin-wrapper your_binary [set of args]
```

Neither wrapper or script are available or being developed now.

**Comment by RalfWerner** [comment]
Ok, I understood. `/system/*` can not be changed when built by `t*.apk` (only Huawei / Gigaset can do that in my case) so there will never be `/bin/bash`. The _'termux-bin-wrapper'_ from your description does not exist. Is this _'termux-package-binary'_ that @fornwall spoke of?

Probably it will never be developed because you consider the _"proot solution"_ to be "future proof" or should `t*.apk` be replaced by `alpine-term.apk` (then I would have more questions)?

**Comment by xeffyr** [comment]
> Is this 'termux-package-binary' that @fornwall spoke of?

Yes, it is.

> or should t*.apk be replaced by alpine-term.apk

alpine-term's parts can be used to execute single binaries https://github.com/xeffyr/alpine-term/blob/master/alpine-term-app/app/src/main/java/alpine/term/TerminalService.java#L319 but whole application cannot.

**Comment by RalfWerner** [comment]
Stupid question about Q/10/SDK 29: If I wanted to update my Gigaset is missing the _"manufacturer update"_ and in my Huawei the termux app would **not** work (many others may not) - _**right?**_
> If you update your device to Android 10, Termux wants work because it uses older sdk.

Ok thanks - so there's no need to adapt termux to the restictions as SDK<29 apps will also work in future releases. So it's just about hypothetical extensions.

My example contains only _**one** single binary_ `s=~/check.exe` but a lot of environment. Do you think a test with _alpine-term.apk_ **could be successful?**

**Comment by xeffyr** [comment]
> Do you think a test with alpine-term.apk could be successful?

Can you explain what exactly you want to achieve:
* Embed binary in APK so it will be executed when this APK opened ?
* Or execute it with alpine-term ?

Solutions for these 2 cases are different and not comparable.

**Comment by RalfWerner** [comment]
Obviously my idea of ​​the SDK=29 adjustment was wrong, since termux works without the _binary restiktionen_ as before on Q-devices. The problem only occurs when the _APK-built-process_ with version 29 is done - _**right?**_

So comparable to the _android-5 branch_, this only becomes relevant when the _android-7-9 branch_ is discontinued or if _Q extensions_ are to be used by Termux. Both are _**not urgent**_, especially since a _termux-sdk=29-version_ (maybe **1.00**) does not exist yet and _Q-Update_ is not yet possible.

Also wrong was my idea of ​​the PATH to _**native lib**_. If not `/system` (I used in the example) _**then where?**_
> Can you explain what exactly you want to achieve:

All in all, _be compatible_ with future devices and apps, and second, if possible, _fix any current issues_ with it.
In my example two APK with the same signing (my-*) are used. I built the `my-x11.apk` (plugin from @twaik) with _your help_ and `*SdkVersion 28` on my PC. It contains at least two bugs (SSM and unexpected _crash_ in direct mode) and currently also the described _**mpv-x bug**_ (also occurs exactly in Xvnc).

I do not know if one of your _two solution suggestions_ can fix the bug - probably not - and which of the two would be used in _termux-**1.00?**_

**Comment by xeffyr** [comment]
> The problem only occurs when the APK-built-process with version 29 is done - right?

Right.

> So comparable to the android-5 branch, this only becomes relevant when the android-7-9 branch is discontinued or if Q extensions are to be used by Termux.

Nothing to do with packages ! Android 7 branch will be used for Android 10.

Problem comes from https://github.com/termux/termux-app/blob/master/app/build.gradle#L18 being set to 29. We will **have** to do this, otherwise Google will delete app from Play market.

Post https://github.com/termux/termux-app/issues/1072#issue-422258684 has the most clear explanation about the problem.

> If not /system (I used in the example) then where?

`/data/app/package.name-${random characters}/lib`

> and which of the two would be used in termux-1.00?

See https://github.com/termux/termux-packages/wiki/Termux-and-Android-10. Embedding in apk has 2 unsolved problems. Until these problems will not be solved, no work on porting Termux to Android 10 will be done.

**Comment by RalfWerner** [comment]
Thanks, my _own APK_ are not affected because they are not offered by _google play_. On the _native lib_:
> `/data/app/package.name-${random characters}/lib` - Embedding in apk has 2 unsolved problems...

I could **not** access also proot (with termux-chroot `ls /data ../data`) because:
`drwxrwx - x 55 system system 4096 Nov 25 13:45 data/` belongs _**system**_. Maybe if I knew `package.name-${random characters}` or have to install _**additional APK / package?**_

to the _**mpv-x**_ bug I have in [dev](https://gitter.im/termux/dev?at=5ddbcc4e4517d002b2eda909) still made a further test.

**Comment by xeffyr** [comment]
> Maybe if I knew package.name-${random characters} or have to install additional APK / package?

```
echo $(dirname $(pm path com.termux | cut -d: -f2))/lib
```

**Comment by RalfWerner** [comment]
yes there is `*/lib/arm64` with two `*.so` (aarch64) and _no_ write permition except _system_.
On Gigaset (arm) I get only `. . . ./lib` als PATH to _**native lib**_ and `echo $(pm path com.termux)` ended with `cmd: Failure calling service package: Failed transaction (2147483646)`. _**You got it too?**_

Embedding `~/check.exe` would require write permission (your first suggestion above?) but do **not** _fix **any of my** current issues above_ - maybe later, when _termux-**1.00**_ exist! Thanks so far.

**Comment by xeffyr** [comment]
> You got it too?

No. But I expected that. The normal Android way of obtaining that path is from java code only.

> Embedding ~/check.exe would require write permission (your first suggestion above?)

No.
If `check.exe` is embedded in APK, it will be extracted to `./lib` directory like other `*.so` files.

**Comment by RalfWerner** [comment]
From the dev team chat yesterday, I understood that the _**./lib**_ - way is currently _not- tracked but **proot**. You have created an `*.apk`, which I could unfortunately **not install** and the attempt of re-signing failed:
```
[~/sd/ip/x] $ apksigner app-debug.apk keystore my-t84.apk
Exception in thread "main" java.io.IOException: not a JavaKeyStore
        at net.fornwall.apksigner.JKS.engineLoad (JKS.java:346)
        at java.security.KeyStore.load (KeyStore.java:1484)
        at net.fornwall.apksigner.KeyStoreFileManager.loadKeyStore (KeyStoreFileManager.java:40)
        at net.fornwall.apksigner.Main.main (Main.java:84)
Exit due to uncaughtException in main thread:
```
You could save the 64 Mb in Github - I did **not** succeed because of the 25Mb limitation. How did you _**do this?**_ In your `*gz` I discovered _SdkVersion 29_ in `*/app/build.gradle`. The X11 plugin of my example was created with _28_ and currently works with `my-t82.apk`. @xeffyr could you please help:
What _**steps should I take**_ to be able to work with **your** `my-t84.apk` and what can be **done _'on device'_?**

**Comment by xeffyr** [comment]
> Should/could the actions tested here above also work on an Andriod-Q device?

From my tests proot'ed Termux appears to work without issues on Android 10. There maybe problems with things like Termux:API but they will be solved later.

Performance is slower under proot by 30-40 % (mostly affects only I/O operations).

You will not see differences in usage experience, binaries can be executed in $HOME in same way as pre-Q Android versions.

> Would that be possible without the proot process (only t84 APK with 29)?

Nothing. Proot acts as executable loader.

> All *.exe are in$HOME like in t82 - isn't that a problem?

Not a problem while proot is being used.

> What can be the reason that sftp does not work anymore after t84 (ssh does it)?

I don't know. This needs testing. But note that proot may not be stable on some devices.

> Is there anything to improve the performance of proot processes (as in t82)?

No. Proot uses `ptrace()` to hook system calls and manipulate their arguments.

You should not execute `termux-chroot` under proot'ed Termux.

**Comment by RalfWerner** [comment]
Pleased I discovered @twaik back in _dev dialog_ with fresh ideas.
Thank you @xeffyr for the explanations (above) I continue here the test. to:
> Performance is slower underproper by 30-40% (mostly Affected only I / O operations).

I understood your answer as saying that _proot_ process _libproot.so_ (`r='pgrep -n libproot'`) is already active in your _t84_, even though _**~**_ is not `/home` but `t=/data/data/com.termux/files; $t/home`, while at _t82_ the _proot_ process has to be created with termux-chroot in order to have comparable proot tests. In addition, the Q-restriction _only_ is bypassed with `ptrace()` and `*.exe` must be saved as before in `$t/*` (sd/ does _not_ works) - _**as far as right?**_

With 30-40% loss, I would have no problem. Now I have the command: `date; you -hd 1 u /; date` without/with _proot_ tested (seconds) and got: _t82_=7/20 and _t84_=25/374 '. With respect to 7, 40% would actually be 7*1.4=_**10 seconds**_. My _unacceptable_ test above result with `check.c` above corresponds to 7/374 although there is hardly any I/O except for _make_. Now you said:
> You should **not** execute `termux-chroot` under proot'ed termux.

Remains a loss factor of 2 to 3 compared to 7 sec. I have described this as justifiable but have not made any more extensive tests and you said the _plugins_ could be a problem. I want to do that below.
## Navigator, plugins: widget, API and two X11 (Xwayland)
All scripts are included in `.bashrc` [sd/ip/s](https://github.com/RalfWerner/integrated-process/blob/master/s) (**134** sloc expanded by `$r` administration). A _light_ version is [api29/t](https://github.com/RalfWerner/integrated-process/blob/master/api29/t) (19 sloc reduced to OTS and _one_ XClient). `t=bash .bashrc $@` is used to start all PID4 processes.
The processes on my _t84_ device are unstable (SSM only here), but they _were already before_ in contrast to the _t82_ device (currently working for 11 days and PID3 since 2 days).
- Each _NEW SESSION_ creates another instance of `$r` (_t84_) and loads/source .bashrc (**28-41**)
- The Widget [GUI](https://github.com/twaik/termux-x11/issues/1#issuecomment-532817407) ff (11 shortcuts) is unchanged (creates additional instances in the Navigator). For _t82_, the shortcut functions are executed as described (**22-25**) (_without_ .bashrc). At _t84_, another libproot.so process `$r` is started and `.bashrc` sourced but not the rest of the shortcut script. If the bash is terminated with `exit` in _t84_, it is also removed from the navigator (without executing the function) with `^d` the bash remains, `kill -9 $r` _not done_, but can no longer be edited.
- API with Widget is only possible on _t82_ (see above) and on _both_ devices, the function can be started with the command: `t d` (**66-87**). It tests most - not just the _11 dialog_ functions of the API plugin (`*sensor` in preparation).
I started the functions (`date; t d; date`) with the 3. (speech) and performed successively until 11. (counter) and finished with 2: (selfie **125-135**).
This has worked on both devices with _t84: _ 141 and _t82: _ 24 seconds. So here is a _loss factor_ of almost **6.** Since the API functions are processed independently of the script, there are unwanted effects on how acoustic prompts "something to speak" sometimes appear after recording and by turning the mobile phone is a longer break at _t84_ before the dialogue appears.
1489/5000
- X11 clients (PID4) are initialized in OTS. With _t84_ the _sleep time_ is often not sufficient to initialize all processes (PID1-4) and almost always at _t82_. For Widget launch - only _t82_ see above. The _pcmanfm Xorg_ shortcut corresponds to the command: `t: 0 x`, adds missing processes from OTS and opens _Xterm_ and _pcmanfm_ in WM _fluxbox_. With the alias `ob` _openbox_ and` fb` can be "toggled" between the two WMs with **loss factor 3** at _t84_ (one second at _t82_). My favorite is _no WM_ with alias `o` - only `kill $WM` and very fast on both devices.
As a last test, I compared _check.c_ (see above) to _view_ of a somewhat larger source file with almost 100 tloc and also included Quick-Edit. The process is subdivided into the actions: start, open, rotate, scroll and change the screen/font sizes. _check.c_ only: circular motion of a graphic on the screen with one finger. In both termux versions, actions are completed after a fraction of a second the longest time it took to "open" _t84_ (about 0.5 seconds). All other actions were about the same in the two versions. The "circling motion" has a marginal "caster" in which the pointer briefly becomes visible under the finger but not different in _t84_.
For Quick-Edit takes: "start" 3 seconds, "open" 26/30 seconds and all other actions are comparable with termux. However, this app has syntax highlighting and advertising (check.c not).
- IP (X11) - can be used for longer autonomous control processes (5 minutes) - in preparation.

**ssh** seems only possible if `$r` _and_ `sshd` are found. Unfortunately, the `pstree` function ends
with:`pstree: error opening uptime file` (**_why?_**) in both versions, but it is also recognizable with `pp` that all `$r` instances with PPID refer to the first one or to _PID1_ (`com .termux` see shots above)

_**Are you interested in further tests?**_

**Comment by xeffyr** [comment]
> Pleased I discovered @twaik back in dev dialog with fresh ideas.

Many ideas that's great. There only one problem - who will implement & maintain that ? So far I haven't seen any working proof-of-concept.

I'd rather go with lower performance (qemu-system or proot) than with regressions (which may take ages to be solved) bringing Termux back to its early stages.

> Unfortunately, the pstree function ends with:pstree: error opening uptime file (why?)

That's SELinux. Since Android SDK does not mention `/proc` as "official" interface, manufacturers are allowed to restrict its access as much as possible (+ for security as /proc has potential for leaking data).

> The pstree problem above seems to be more complicated. by replacement with cd u /; p = pstree; mv $ p - $ p; In -s busybox $ p is the function again available (error in coreutil?)

Not an error. Just different implementation.

>  Due to various sevices and libproot this is no longer possible and also alias k = kill $PPID from .bashrc is no longer possible in t84 - option -9 is required.

libproot.so is the $PPID now. Proot can't be safely terminated as immediately destroys all child processes.

**Comment by RalfWerner** [comment]
Thanks @xeffyr for your answer, a few questions are still open.
> I'd rather go with **lower performance** (qemu system or **proot**) than with regression (which may take **ages** to be solved) bringing Termux back to its early stages.

API loss factor of _**6 times of t82**_ is a bit too much - _**can this be improved?**_ I agree with you to avoid regression. So _pstree_ in busybox is a _working_ alternative (like _ps_ before). If @twaik had a _working_ test environment in a **manageable time** with his ideas, I would like to test them as well.
> libproot.so is the $PPID now.

Unfortunately in my test _t84_ often came in a mode that only allows _restart termux and services_. So I've included `kill -9 $PPID` in the`alias k`.
Widget is _**no longer usable**_ at _t84/proot2_ (above)! Could you find out why _Widget/bash_ behaves differently on _t82_ and _t84_ and if this _**can be fixed?**_

**Comment by xeffyr** [comment]
> can this be improved?

No. Proot intercepts all system calls and changes their arguments, that is core functionality.

> Widget is no longer usable at t84/proot2 (above)!

My APK is just example with using proot to launch shell. I'm not aware about any compatibility with plugins.

**Comment by RalfWerner** [comment]
>I maybe will submit newer version soon

That's good - please let me know when the time comes. Once again to one widget (.shortcut): `.shortcuts/ls-env-ps`. in _**t82**_ execute _without_ a new _bash_ the included script:
`ls -la; env; ps -eo pid, ppid, comm, etime; pstree; exit 1`  with a new navigator item, like also if the script is executable in both versions. _t84_ starts a bash (11745) - **without script** and generates and holds also a new navigator item. The following processes are active:
```
31767[~]$ .shortcuts/ls-env-ps
...
XDG_RUNTIME_DIR=u/tmp
SSH_CLIENT=192.168.0.114 64919 8022
PATH=./:/data/data/com.termux/files/home/:./:/data/data/com.termux/files/home/:/data/data/com.termux/files/usr/bin:/data/data/com.termux/files/usr/bin/applets
SSH_TTY=/dev/pts/1
_=/data/data/com.termux/files/usr/bin/env
PID   PPID  COMMAND          ELAPSED
 2171 32503 bash              1h41
 2173  2171 bash              1h41
 2219  2173 pcmanfm           1h41
 2229  2173 aterm             1h41
 2236  2229 bash              1h41
 5376 32503 bash              1h02
 5377  5376 bash              1h02
 5404  5377 aterm             1h02
 5405  5404 bash              1h02
 5874 32503 openbox          58:40
11442 32503 bash              0:02
11445 11442 ps                0:01
31653   529 com.termux        1h44
31767 31653 libproot.so       1h43
31773 31767 bash              1h43
31842     1 sshd              1h43
31970     1 Xvnc              1h43
32158 31773 Xwayland          1h43
32501 31842 sshd              1h43
32503 32501 bash              1h43
?-+-Xvnc
  `-sshd---sshd---bash-+-bash---bash-+-aterm---bash
                       |             `-pcmanfm-+-{gmain}
                       |                       `-{pcmanfm}
                       |-bash---bash---aterm---bash
                       |-bash---pstree
                       `-openbox
31767[~]$
31767[~]$ cat .shortcuts/ls-env-ps
ls -la;env;ps -eo pid,ppid,comm,etime;pstree;exit 1

--- Here I have started the Widget from start screen. ---

31767[~]$ pp
PID   PPID  COMMAND          ELAPSED
 2171 32503 bash              1h54
 2173  2171 bash              1h54
 2219  2173 pcmanfm           1h54
 2229  2173 aterm             1h54
 2236  2229 bash              1h54
 5376 32503 bash              1h15
 5377  5376 bash              1h15
 5404  5377 aterm             1h15
 5405  5404 bash              1h15
 5874 32503 openbox           1h11
11745 31653 libproot.so       2:16
11751 11745 bash              2:16
12066 32503 ps                0:01
31653   529 com.termux        1h56
31767 31653 libproot.so       1h56
31773 31767 bash              1h56
31842     1 sshd              1h56
31970     1 Xvnc              1h56
32158 31773 Xwayland          1h56
32501 31842 sshd              1h56
32503 32501 bash              1h56
```
Can I do something to _**start the script here?**_

**Comment by RalfWerner** [comment]
I did a few tests with [Ubuntu](https://wiki.termux.com/wiki/Ubuntu) to possibly fix / work around some of the Xserver bugs. I first installed the packages _xterm_ and _nedit_ there. With the _DISPLAY_ of my PC (Xserver) these also worked fine.
After the installation I start _Ubuntu_ (root) with the alias `uu`, on:
- _t82_ in **~/uu** with `-b/sdcard`. Logging in is not a problem here and the symlinks: _/sdcard sd, sd/ip i, ../usr u_ and _i/s .bashrc_ were created as in termux, so that root behaves like termux, but only the functions _~/u_, here _uu/ubuntu-fs/usr/_ are valid here.
- _t84_ in **sd/uu** (start-ubuntu.sh) calls proot with parameters. This should probably be _not_ done, since I am already in proot.

What should the command _uu in **t84**_ and its parameters look like here and how could I _**use the X11/server**_ (PID2+3=Xwayland) _** there (proot2)?**_

**Comment by xeffyr** [comment]
> This should probably be not done, since I am already in proot

This should be done to call proot with different parameters. Proot in my APK is not meant for just `execve` workaround but for changing file system layout (pending for next prooted Termux version) so we will be able to make Termux prefix-independent and also workaround some filesystem problems in Android 10.

> What should the command uu in t84 and its parameters look like here

Normally parameters should be same.
Regarding nested proot - nothing can be done as reusing existing proot process is not possible.

> and how could I use the X11/server (PID2+3=Xwayland) there (proot2)?

Probably `$PREFIX/tmp` should be mounted (with `-b`) to `/tmp`.

**Comment by RalfWerner** [comment]
## Thank you - I still have a few questions (?) to be successful.
In order to use my _make loops_ in _**uu**_ (see below), several development packages have to be installed. That takes about 800Mb in `u*-fs` and [script s](https://github.com/RalfWerner/integrated-process/blob/master/s) (OTS in preparation) is adapted to this location (root).

For _t82,_ I was able to start the _make loop_ `t :0 c` (shot4 in the first commit above), but do not get any output in PID2 and `*.exe` and aborted at `dpy=XOpenDisplay (0)`. I suspect an Xwayland problem (`could not connect ..` below). _pp_ shows the processes in termux. The PID2 process (4037) is only visible with wtermux with `my-x11.apk` the _service_ in _termux is _not_ visible and _**cannot** be started from ubuntu - or **do you've an idea?**_
```
u=uu/uu.sh; echo '#!/data/data/com.termux/files/usr/bin/bash
cd $(dirname $0); unset LD_PRELOAD
command="proot --link2symlink -0 -r ubuntu-fs -b $PREFIX/tmp:/tmp -b /dev -b /proc -b /sdcard -w /root /usr/bin/env -i HOME=/root PATH=/usr/local/sbin:/usr/local/bin:/bin:/usr/bin:/sbin:/usr/sbin "
command+="TERM=xterm LANG=C.UTF-8 /bin/bash --login"
if [ -z "$1" ];then exec $command; else $command -c $@; fi'>$u; chmod +x $u; alias uu=$u
[~]$ uu
...
moin IP 192.168.0.132, Xwayland:2002, Xvnc:13094
root[~]# t :0 c
make: '/root/TMP/check.exe' is up to date.
:0 (c) /root/TMP/check.exe P*pm s *.c pet1*m :182834.400
E=> Display not opened! STOP
root[~]# Xwayland :0&
[1] 3984
root[~]# could not connect to wayland server
(EE) Fatal server error:
(EE) Couldn't add screen

[~]$ pp
PID   PPID  COMMAND          ELAPSED
 2000     1 bash              2d02
 2002  2000 Xwayland          2d02
 3888 29085 sshd             10:19
 3889 29085 sshd             10:16
 3892  3888 bash             10:15
 3924  3892 proot             8:01
 3927  3924 bash              8:01
 4002 29085 sshd              0:19
 4007 29085 sshd              0:10
 4008  4002 bash              0:10
 4037   322 .termux.wtermux   0:06
 4059  4008 ps                0:01
13094     1 Xvnc              1d01
24407  9231 bash              9h21
29085     1 sshd              2d05
```
>nested proot -nothing can be done as reusing existing proot process is not possible.

I had already tested and excluded the proot start from _t84_ above - or mostly you _**something else?**_
To decouple the `u*-fs` from the termux installation I want to use` -r $HOME/sd/uu/ubuntu-fs -b $HOME/sd:/sdcard` - _**is that a problem?**_
```
20258[~]$ bash sd/uu/uu.sh
proot error: execve("/usr/bin/env"): Permission denied
proot info: possible causes:
  * the program is a script but its interpreter (eg. /bin/sh) was not found;
  * the program is an ELF but its interpreter (eg. ld-linux.so) was not found;
  * the program is a foreign binary but qemu was not specified;
  * qemu does not work correctly (if specified);
  * the loader was not found or doesn't work.
fatal error: see `proot --help`.
proot error: can't chmod '/data/data/com.termux/files/usr/tmp/proot-19838-VykanY': No such file or directory
```

**Comment by RalfWerner** [comment]
@xeffyr test for [termux / ubuntu] (https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-572692683) on aarch64 (t84 with proot2):
- `apt-get update` is required (not for t82) and takes 6 minutes afterwards` apt-get install xterm` is possible and takes 20 minutes.
- Then I opened an xterm from termux / ubuntu on the PC-Xserver and listed the data from `/sdcard/*` with `ll sd/*`, which took 0.5/15 seconds (with t82, both need significantly less than one second).

The t84 behavior is unacceptable and I gave up starting an ubuntu Xserver there. With t82 X11 server does not work and I could not use the Xvnc server (PID3 = 13094 from OTS with tigervnc in termux) in ubuntu. If I kill the process before _uu_, four new processes will be created with _vncserver_, but they can be killed when RealVNC starts You can see why? Processes and log:
```
[~]$ pp
PID   PPID  COMMAND          ELAPSED
13094     1 Xvnc              3d21
[~]$ pp
16393     1 Xtigervnc         1:16
16394     1 vncserver         1:16
16395 16394 Xvnc-session      1:16
16396 16395 vncconfig         1:16

root[~]# cat .vnc/localhost\:1.log
Xvnc TigerVNC 1.9.0 - built Dec  1 2018 21:51:29
Copyright (C) 1999-2018 TigerVNC Team and many others (see README.rst)
See http://www.tigervnc.org for information on TigerVNC.
Underlying X server release 12003000, The X.Org Foundation

Sun Jan 12 14:42:52 2020
 vncext:      VNC extension running!
 vncext:      Listening for VNC connections on local interface(s), port 5901
 vncext:      created VNC server for screen 0

Sun Jan 12 15:01:02 2020
 Connections: accepted: 127.0.0.1::48058
 SConnection: Client needs protocol version 3.8
 SConnection: Client requests security type VncAuth(2)
terminate called after throwing an instance of 'rdr::Exception'
terminate called recursively
(EE) Backtrace:
(EE) Fatal server error:
(EE) Caught signal 6 (Aborted). Server aborting
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 173 requests (173 known processed) with 0 events remaining.
Killing Xtigervnc process ID 16393... which was already dead
Cleaning stale pidfile '/root/.vnc/localhost:1.pid'!
```

**Comment by RalfWerner** [comment]
## Xvnc in proot/ubuntu
@xeffyr Your [log](https://linx.li/zz7hejl2.txt) handles _**pu**_ in ubuntu and _**PID3**=vncserver_ from OTS. _**Not** recognizable_ are the four processes there (16393 ff above)
Here at _Xvnc (:1) **PID2**=RealVNC-App_ can be called before or after PID3 start (only if it does not yet exist). Since it already exists through **PID1** (termux app start with OTS), I get a _Open-Display-Error_ at _**PID4**=check.exe_ in ubuntu and _without Xvnc_ the `XIO: fatal IO error` at the end,
_as **you** get too and how to use _**PID3**=Xwayland_ (:0) **what is to do?**_ 
## t64.apk instead of t84.apk
I gave up the test of your _proot2.apk_ due to serious defects (no widget and loss factor of 80 times at proot) and t84 by your apk [above](https://gitter.im/termux/dev?at=5e1b790f6f604152992cdc5b) (my-t64 with apksigner) and describe your suggestions in [README](https://github.com/RalfWerner/integrated-process/blob/master/api29).
As the first steps I did _**tu**=t.tgz update_ and _**pu**=apt-get ... (ubuntu)_. Pleasing: _Widget and sftp_ work again and _**pu**_ is ready after 5 minutes (960Mb). Unfortunately, your links are replaced after _**tu**_. The `$NL/arm64/*| grep -c lib_` 76 files and `...|grep -c bin_` 149 files are still there (38Mb) and you donated a "brilliant" script for repetition - will experiment a little more.
`NL=$(dirname $(pm path com.termux|cut -d: -f2))/lib` works well for t64 but _how to do it for **t32?**_

**Comment by RalfWerner** [comment]
@xeffyr Your [log](https://linx.li/zz7hejl2.txt) handles _**pu**_ in ubuntu and _**PID3**=vncserver_ from OTS but ends with the same `XIO: fatal IO error` as I'ver got **and** I gave up the test of your _proot2.apk_ due to serious defects (no widget and loss factor of 80 times at proot) and replace t84 by your apk [above](https://gitter.im/termux/dev?at=5e1b790f6f604152992cdc5b) (t64) [more](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-574123255)

**Comment by xeffyr** [comment]
> how to use PID3=Xwayland (:0) what is to do?

Either to launch it in TCP mode or bind-mount (with proot) $PREFIX/tmp directory to /tmp.

> NL=$(dirname $(pm path com.termux|cut -d: -f2))/lib works well for t64 but how to do it for t32?

If I correctly understand, **t32** is Termux for 32 bit ARM. I currently have APK only supporting AArch64 and i686.

> The first hurdle is the symlinks in sd (which is not a problem from termux), which you said "could do" 

Which context you are referring to ? I didn't suggested to put environment on /sdcard.

**Comment by RalfWerner** [comment]
>Either to launch it in TCP mode or bind-mount (with proot) _**$PREFIX/tmp**_ directory to /tmp.

**Done!** (find backwards `$PREFIX/tmp` in this issue and checked with `ll /tmp`. With Xvnc _**you**_ also had _**no success**_ (log). Have you been able to start one of the two Xservers (preferably both) _**on your device?**_

Unfortunately, my third Xserver (XSDL) now works on neither of my two devices after the typical _"6-second-init"_ shows the following message (both devices)):
![grafik](https://user-images.githubusercontent.com/45426380/72418334-ea05ae00-377a-11ea-8c26-e93bd847bcae.png)
This is still the case even after _"reboot without termux"_. Trying _de/re-install_ the app is only possible on t82 and on t64 it is no longer possible to _"reinstall"_ (google-play) _**did you have that too?**_ I did this with `nedit -display ...` (shot2) and it works well on my PC (with the Windows Xserver). The functions _ps_ and _ip_ (wrapper on /system/bin/ip) are still missing on _**uu**_. with `cp u/bin/ip uu/ubuntu-fs/bin` (termux),` -b system` (uu.sh) and `#!/bin/sh` (in /bin/ip) I get `/bin/ip: 3: exec: /system/bin/ip: not found` - 
_**what is wrong?**_
>I currently have APK only supporting AArch64 ... Which context you are referring to (**sd symlinks**)?

t64 is sufficient for check now! ... Unfortunately I couldn't find the reference anymore - no matter if you say _"it never works"_ I believe you. I am interested in storing data in _**sd**_ in order to be _independent_ from the de/re-installation of termux as with `t.tgz` - to _"disposed of beforehand" **you've said nothing!?**_

**Comment by xeffyr** [comment]
> Have you been able to start one of the two Xservers (preferably both) on your device?

XSDL works fine. No special setup needed, did `export DISPLAY=:0` and it worked.

Using GIMP in Termux/Ubuntu:
![Screenshot_20200115-142024_XServer XSDL](https://user-images.githubusercontent.com/25881154/72433695-97d88300-37a2-11ea-8576-d679f3d0ccae.jpg)

However there are SHM-related errors which cause certain issues. Example:
```
shmget() failed: Function not implemented
Disabling shared memory tile transportgimp_get_monitor_resolution(): GDK returned bogus values for the monitor resolution, using 96 dpi instead.
```
Proot does not implement system calls like `shmget` and makes them no-op. This does not affect programs from Termux environment as here SHM functionality is implemented through libandroid-shmem.so & Android's ashmem.

> Trying de/re-install the app is only possible on t82 and on t64 it is no longer possible to "reinstall" (google-play) did you have that too?

No. I have reinstalled Termux multiple times recently to switch between versions, no issues regarding this observed.

Of course all apps & addons signed by one key.

> what is wrong @xeffyr? - sorry forgot to restart uu but ps still missing!

`-b system` is wrong and will lead to following issue
```
proot warning: can't sanitize binding "system": No such file or directory
```
Use `-b /system` instead.

> I am interested in storing data in sd in order to be independent from the de/re-installation of termux as with t.tgz - to "disposed of beforehand" you've said nothing!?

`/sdcard` support will be dropped for API 29 and higher as Android no longer provides access at file system level. Termux will not support to have stuff on /sdcard for this reason and few others.

Binary execution from sdcard is possible despite noexec mount option, just proot built from the branch https://github.com/termux/proot/tree/force-exec is needed.

Storing of symlinks and other special files (directly) is not possible due to file system implementation limits.

* /storage/emulated/0 (/sdcard) is technically FUSE-based overlay for /data/media/0 located on EXT4 (sometimes F2FS) partition. It is used to mimic file system like FAT32 or exFAT while still providing basic access control through ownership:
   ```
   herolte:/sdcard $ mount | grep /storage/emulated                               
   /data/media on /storage/emulated type sdcardfs (rw,nosuid,nodev,noexec,noatime,fsuid=1023,fsgid=1023,gid=1015,multiuser,mask=6,derive_gid,reserved=20MB)
   herolte:/sdcard $ ls -ld Android/data/x.org.server                             
   drwxrwx--x 3 u0_a725 sdcard_rw 4096 2020-01-15 14:05 Android/data/x.org.server
   herolte:/sdcard $ ls -ld Android                                               
   drwxrwx--x 6 root sdcard_rw 4096 2019-12-29 16:36 Android
   herolte:/sdcard $ ls -ld .                                                     
   drwxrwx--x 15 root sdcard_rw 4096 2020-01-15 05:01 .
   ```
* File type limits can be avoided by storing data inside file system image. But this is possible if device is rooted (image stored on internal or external sdcard or even USB stick and mounted with `mount`). Another variant (no root required) will be use of QEMU in system mode but running environment in this way is out of Termux scope currently (due to performance & usage complexity issues).

**Comment by RalfWerner** [comment]
> XSDL works fine (... been able to start one of the two Xservers ...)

@xeffyr Thank you first! I meant Xwayland/X11 and Xvnc! XSDL is only my 3rd choice and X11 is the 1st - you apparently haven't _**had any success with that yet - have you?**_
With a workaround (download/installation from an external url) I now have XSDL on _t64_. The problem was my _"play store app"_ (not your beer) but - surprise - the app works again overnight. Fdroid, no longer offers XSDL but termux (84). Currently, however, I only use _**my** self-signed apk_ (t64).
> /sdcard support will be **dropped for API29** ... But this is possible if **device is rooted**

With `sd=/sdcard; s1=/storage/6533-6333; s2=/storage/emulated/0` I have three possible paths that start with `[ ! -d $sd ]&& sd=$s1; [ ! -d $sd ]&& sd=$s2` of the device environment. Then in _$HOME_ e.g. set up symlinks with `ln -s $sd/api29 i` and `ln -s i/t .bashrc`, so that termux and ubuntu use the same scripts but have different _/home_ paths.
_**This won't work with API29 - correct?**_
Even saving/resetting (t.tgz - _tu_) to _$sd_ is _**no longer possible**_ and with every termux de/re-install the complete function set must be installed with _pu_ (very time-consuming).
I installed mpv in _uu_ again (one hour) and was a little more successful than above (find mpv backwards here). The function shows the video (-vo), but gives no sound (-ao reversed as with termux), many error messages and just only possible on XSDL. The app is then 3.15Gb and $s2 is almost full - _**unacceptable!**_
You say: _if device is rooted_ all problems should be solved. How would I do that and is the cell phone still usable apart from _temux_?
> Use -b /system instead.

Typo `-b system` was only in this issue (correct in the script) what was missing was: _crtl-d_ plus _uu_ (restart to activate). then I added -b /data and NL=$NL to uu.sh, moved it to `*/i/uu.sh`, use`alias uu="bash i/uu.sh" `and preparing some widgets /.shortcuts with `nedit -display ...` (shots in preparation) and also on my PC and `mpv` wrapper.
I completely removed the functions _ps_ and _ip_ in _**uu**_ and added _/system/bin_ in _PATH_. This means that both functions (for process and display control) are correct.
Then I am currently working on your _ELF-analyze-loop_ (/bin, /lib) in `i/uu.sh`, so that I can also analyze actual packages and _**my** make loop. Are you **interested in the result?**_

**Comment by xeffyr** [comment]
> had any success with that yet - have you?

**Comment by xeffyr** [comment]
> had any success with that yet - have you?

I don't use Xwayland. May try with Xvnc though.

> This won't work with API29 - correct?

You specified these paths: `sd=/sdcard; s1=/storage/6533-6333; s2=/storage/emulated/0`:

* /sdcard  - a compatibility symlink to shared storage of primary device account:
   ```
   herolte:/ $ realpath /sdcard
   /storage/emulated/0
   ```
* /storage/6533-6333  - external sdcard or usb drive
* /storage/emulated/0  - shared storage for primary device account (/sdcard is an alias to this one).

Both `/storage/emulated/0` and `/storage/6533-6333` do not allow file execution. Additionally, API 29 [disallows](https://android-review.googlesource.com/c/platform/system/sepolicy/+/804149/5/private/app_neverallows.te) execution from `/data/data/*`. 

Should I mention that **you can't put environment to these locations** ? Perhaps should be obviously that it will not work (it will be even broken due to nonexistent symlinks).

If you want to put environment on /sdcard or its equivalent, you can do this only through EXT4 file system image. Example command sequence for ADB root shell:
```
setenforce 0
dd if=/dev/zero of=/sdcard/data.img bs=1M count=2048
mke2fs -t ext4 /sdcard/data.img
mkdir /mnt/termux-env
mount /sdcard/data.img /mnt/termux-env
cp -a /data/data/com.termux/files/* /mnt/termux-env/
rm -rf /data/data/com.termux/files/*
ln -s /mnt/termux-env/home /data/data/com.termux/files/home
ln -s /mnt/termux-env/usr /data/data/com.termux/files/usr
```

> The app is then 3.15Gb and $s2 is almost full - unacceptable!

New Termux with API 29 fix will take **2x** more space (size of: apk + extracted data to /data/app/com.termux-*/lib).

> if device is rooted all problems should be solved.

Rooting by itself doesn't solve problem. Additional actions will be required.

You will be able to turn off SELinux restrictions with `setenforce 0` and be able to execute stuff in `/data/data/com.termux/files/home/*`.

> How would I do that and is the cell phone still usable apart from temux?

See commands for ADB above.

**Comment by RalfWerner** [comment]
> / sdcard - a compatibility symlink to shared storage ... do _not allow file execution_

only on _t82_ and **not** your _t64_ (my Huawei device) - that's why I check the existence. I have had a symlink on /storage/sdcard1 since 1970 /sdcard - which doesn't exist but $s1. Have I _**misunderstood**_ that data access (`ln -s $sd...`) _**is also possible in API29?**_
_no execution_ has always been this way and I have aligned my processes accordingly: `. .bashrc` is passed _only **sourced**_ and `i/uu.sh` as argument from `u/bin/bash`. If this is still the case in API29, I am reassured.
What changes is then _ only_ that _ubin / bash_ is a link to `$NL/*bash.so` (/data/app...)
that **you** created on Aug 18 11:54. _**Do I understand it correctly now?**_
> New Termux with API 29 fix will take 2x more space

see _**Addendum to a ELF-check**_ above!  The APK should be at least _**1441M in size - right?**_
From **my _make-loop_** it is only _one_ file with 110K, which is saved with make in `s=~/check.exe`. With ubuntu then _**2x**_. In the first [comit](https://github.com/RalfWerner/integrated-process/issues/2#issue-524327794) (shot4) the Xvnc display is included. Editing the code there or with _QuickEdit_ (shot2) - requires saving to `$sd`. If you click on the 2nd icon (blue/yellow barrel) in the check-Xclient, `make $s; $s` done. That means after 2-0.5 seconds the screen appears again with updated code and without change after 0.2 seconds. The replacement of `$s=$NL/lib_check_.so` (if I had write permission) would also work. _**How can I achieve this?**_
>Example command sequence for ADB root shell

Thanks @xeffyr for the description I will test it later with a _smaller_ app. With the exception of _setenforce_, the process could also run under termux. however, `pkg install e2fsprogs` would have to be done to have _**mke2fs**_, but _pkg_ no longer works with API29, since `$NL/libbin_mke2fs.so` _cannot_ be generated.
_**How do you free yourself from the dilemma?**_

**Comment by xeffyr** [comment]
> Have I misunderstood that data access (ln -s $sd...) is also possible in API29?

No, not possible unless you will use private app directory on such storage:
```
$sd/Android/data/com.termux
```

> What changes is then _ only_ that ubin / bash is a link to $NL/*bash.so (/data/app...)
that you created on Aug 18 11:54. Do I understand it correctly now?

Generally yes, but we are going to put everything (not just binaries) to `$NL` and then setup symlinks in $PREFIX.

> The APK should be at least 1441M in size - right?

Assume that APK of `perl` takes 22 MB:
```
~/perl-app/app/build/outputs/apk/debug$ du -hs app-debug.apk 
22M	app-debug.apk
```
Size of extracted jni lib directory takes 65 MB:
```
~/perl-app/app/src/main/jniLibs$ du -hs arm64-v8a/
65M	arm64-v8a/
```

What we have when package is installed:
* APK is copied to `/data/app/*` and takes its 22 MB.
* Lib directory is being extracted and takes its 65 MB.
So total space usage will be 87 MB.

Reality can be a bit worse:
![Screenshot_20200117-170334_Settings](https://user-images.githubusercontent.com/25881154/72622301-9a2a1100-394b-11ea-87bf-e84e41b456dd.jpg)

Example with `perl` package is not perfect since it contains a lot of text files which are compressed well inside APK.

> How can I achieve this?

You can execute your `check` under proot.

In discussion at https://gitter.im/termux/dev was suggestion to reimplement `termux-exec` (https://github.com/termux/termux-exec) to prepend `proot` to command line of any userspace executable at time when `execve()` is invoked.

>  however, pkg install e2fsprogs would have to be done to have mke2fs

No need. My example was for ADB shell. All devices have at least `make_ext4fs` which can be used too. Some have `mke2fs` too in /system/bin.

Usage for `/system/bin/make_ext4fs`:
```
make_ext4fs -l 1G /sdcard/test.img /path/to/fs/root/dir
```

> but pkg no longer works with API29

In API 29 `pkg` will become a Termux custom package manager. APT will not be used anymore but its repositories will continue exist.

> and what about t32?

**t32**: https://linx.li/t32-termux-api29.apk

Note that proot-based Termux apk and examples I have posted on Gitter are discontinued -- Fornwall created https://github.com/termux/termux-app/pull/1427 where implementation is better than mine and close to one that will be used. But bootstraps here are not API-29 compatible currently.

APK for PR 1427: https://linx.li/termux-pull1427-debug.apk

> you tell me if you were successful with Xvnc in proot/ubuntu?

Xvnc was fixed with
```
vncserver -SecurityTypes None :1
```

![Screenshot_20200117-180552_VNC Viewer](https://user-images.githubusercontent.com/25881154/72626829-1163a300-3954-11ea-9259-c70a1c392005.jpg)

**Comment by RalfWerner** [comment]
Emulation of Q10 (studio) worked after various de/reinstallations and "windows features" (10Gb) and BlueStacks (api25 with 5.6Gb) after update. A test is therefore only possible in Studio, as there is no Q10 update planned for my devices. The following shot shows the different sources with _t32_ from @xeffyr. 
![grafik](https://user-images.githubusercontent.com/45426380/74610615-ef954180-50f4-11ea-9079-0cc9aa086228.png)
So the APK contains next to armeabi-v7a (32bit) _also lib_ from **x86** (I used it here) and arm64-v8a (should be identical to t64). Unfortunately there is no _**proot*_ in the APK and only the _"classic"_ pkg (replaces the symlinks on nl) - _**do you have/plan a APK with it?**_
like `ln -s ../usr u` I created` ~/nl` (native libary) as symlink - here with:
The lib (u/bin, u/lib* and nl) should be identical before _pkg_ (100% symlinks) and can then be used in BlueStacks as on my devices. `tu.tgz` for backup, if reinstall is required, could be created in /storage/emulated/0 and also exported to studio so that it can be disposed of with the termux deinstallation _not_.

I have little experience with _**studio**_ emulation/AKP. I tested _t73_ (old school) with net-traffic-bootstrap after drag/drop in the emulation. But there is also the way to start _without installation_ termux from _studio - can so **several** termux versions be tested in **parallel?**_
- more later...

**Comment by xeffyr** [comment]
>  do you have/plan a APK with it?

I do not plan any new apks. We already know all properties of Android 10 we need. No more tests from me until some implementation will be already done.

https://github.com/termux/termux-app/pull/1427 - is the current *progress*.

**Comment by RalfWerner** [comment]
> I do not plan any new apks. We already know all properties of Android 10 we need

### conlusion of api29 test (above)
Since with _**t32 no proot**_ function is available in Android Q/10 and obviously you haven't had any success either, I cannot test this _**important**_ function for _make-loops_.
The storage restrictions have been confirmed for _non-google-apps._ These _**must be used**_ to protect themselves against loss and are very complex and time-consuming.

With api28, objects can be _easily installed and exchanged_, even if they contain complex tasks such as OTS - that's the end of it! If you should @xeffyr create an apk _**with proot**_ please let me know :)
I will continue my research ubuntu/Xserver (above) and [at widget](https://github.com/termux/termux-widget/issues/39).

**Comment by RalfWerner** [comment]
## Continue research of ubuntu/Xserver

Currently I need about 45Gb on the PC for three Android emulators and ApkProjects. After the actions in the last commit, I was able to use _*nedit*_ in emulators according to the _**pu**_ from the Q10 emulator, but have no functions like _ps/pp, grep, sed_ that are not available in `*p29.apk` and lead to error messages with scripts.
Shot: R11 emulator with nedit (shows first lines of uu.sh) with the PC-Xserver
![grafik](https://user-images.githubusercontent.com/45426380/75792810-c3252a80-5d6e-11ea-9450-22fe2caad510.png)
### 2nd choice Xserver: Xvnc
The third choice would be XSDL, whose use is limited to the creation of shots (above). The Xvnc server in termux is not identical to the Ubuntu version (uu). Both are installed with tiger*, use different, mutually exclusive processes, but both have the same real viewer (a change must be confirmed there).

With your/@xeffyr [uu-shot above](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-575689025) the WM=fluxbox+Xterm is started with the Xserver. It shows the problems why Xvnc is _**not 1st choice**_: Many XEvents _**not**_ get through to the Xclient but are treated by the viewer as if you were on a PC with mouse+KB. e.g. a simple rotation of the phone does nothing on a 1900x1200 screen (config). The real size is _**not**_ used, but Real-Viewer only tries _poorly_ to compensate for the shortage. Compared to X11/Xwayland, the viewer _**always**_ works in _emulate touchpad mode_ but correct would be _direct mode_, which almost all apps use.

To use XClients according to the device properties (mobile phone), the 1st choice should use (Xorg on DISPLAY=: 0) and _**no**_ WM. Unfortunately I still have no idea how this works in _**ubuntu - you? - please give me a hint!**_

**Comment by godsworkers** [comment]
Is it possible to redirect storage location inside git on termux from remote=/data/data/com.termux/files/.git
Redirected into
/storage/emulated/0/Android/data/com.termux/files
?

**Comment by RalfWerner** [comment]
I have checked for Android 11 and `/storage` [here](https://github.com/termux/termux-app/issues/1617#issuecomment-695179562) and [here](https://gitter.im/termux/dev?at=5f75d782f9cfbe19f930de61). Your `g=.git` is in the path` ../ ` on termux and can be moved (`mv`) to your target path (corresponds to `$s/$a`) and replaced with a symlink (`ln -s ../$g $s/$a/$g`).

**Comment by godsworkers** [comment]
How will you write it ex?

**Comment by RalfWerner** [comment]
> How will you write it ex?

What do you mean by that? Which termux version do you use?
I'm treating _targetSdk>**28**_ here. With _**28**_ `$a` is mostly not necessary, with _**29**_ sometimes, with _**30**_ almost always and also depending on the device.

**Comment by godsworkers** [comment]
> I have checked for Android 11 and `/storage` [here](https://github.com/termux/termux-app/issues/1617#issuecomment-695179562) and [here](https://gitter.im/termux/dev?at=5f75d782f9cfbe19f930de61). Your `g=.git` is in the path` ../ ` on termux and can be moved (`mv`) to your target path (corresponds to `$s/$a`) and replaced with a symlink (`ln -s $s/$a/$g ../$g`).

Termux . 28 from f-droid
And  termux.api_v31.  From the library    https://archive.org/details/com.termux.api_v31

**Comment by RalfWerner** [comment]
Then you are here in the **wrong issue** the command lines (`mv ..., ln ...`) above shouldn't be a problem for you.

**Comment by godsworkers** [comment]
> > I have checked for Android 11 and `/storage` [here](https://github.com/termux/termux-app/issues/1617#issuecomment-695179562) and [here](https://gitter.im/termux/dev?at=5f75d782f9cfbe19f930de61). Your `g=.git` is in the path` ../ ` on termux and can be moved (`mv`) to your target path (corresponds to `$s/$a`) and replaced with a symlink (`ln -s $s/$a/$g ../$g`).
> 
> Termux . 28 from f-droid
> And  termux.api_v31.  From the library    https://archive.org/details/com.termux.api_v31
> 
> 

Setting up the directory for git.  termux is a self fix ( the root files can be seen from others viewing on different device) google is only hiding roots from the device owner but not others

**Comment by RalfWerner** [comment]
Again: This issue _**not**_ dealt with google or fdroid. Find another forum to discuss this!

**Comment by RalfWerner** [comment]
> The **last** _google play store_ version is **101** and `pkg up` (e.g. `liblz4`) is still possible. A termux version with level 29+ in fdroid is currently _not planned_ but possibly newer versions with level 28 there. _**So far correct?**_

@xeffyr I didn't get an answer to _**this**_ question so I assume _**Yes**_, related to [wiki](https://github.com/termux/termux-packages/wiki/Termux-and-Android-10)! Termux with _target level 29+_ is then a _private build_ of `*.apk` - I've done several times. The packages are then in `bootstrap-$a.zip` or/and other _plugins_ but **you** are not serviced _these._
>Without a little background on how Termux is being implemented, it may be hard to **understand**

You have a few other ideas for people who **understand.** in between I count myself among **them** and have bought such a device, but I ask myself, what is the _benefit of target **29+**_, if the developers have given up - _**what do you think?**_

**Comment by xeffyr** [comment]
> I didn't get an answer to this question so I assume Yes, related to wiki!

Yes, Termux will have target SDK level 28 and be published on F-Droid (v0.102+).

> what is the benefit of target 29+

Primary benefit is *to be able publish Termux updates on Play Store* and also *to avoid possibility of delisting application as "deprecated" by Google* from this installation source.

**Comment by RalfWerner** [comment]
> Benefits of **in-APK packaging**

Thanks @xeffyr for the feedback. You repeated the _benefits_ because I didn't ask about _limits_ that led to the development being canceled - OK so far - but _**"in-APK packaging"**_ should be able to work in _28-Termux_ as well as in _**29.**_ The almost daily updates of the package servers would then have to be done _also_ in [apks](http://termux.net/apks/) or _only there_ if not two Termux versions are exist in parallel and all developers adhere to them - _**is that realistic?**_
My practical [experience](https://gitter.im/termux/dev?at=5f205a96e90668200522ae2e) with this concept is rather frustrating and the my most important [29+ limit](https://gitter.im/termux/dev?at=5f75d782f9cfbe19f930de61) unaffected.

I got the first _**29**-Version_ from _you_ in Sept. 2019 (with ELF split).
My current idea is to keep the _apt/pkg_ updates in 29 versions (test without problem) to create new `bootstrap$a.zip` from the _backups_ after `pkg in/up`, which will be used for building the next `*.apk` generation but with ELF-Split and complete _symlink table_ to `nl=../usr/nl`. That would also be possible at 28 and would only require _one_ [symlink](https://github.com/termux/termux-app/issues/1617#issuecomment-695179562) for _**29 - what do you think of that?**_

**Comment by RalfWerner** [comment]
## Device (Pad) with _**PiP**_ (Picture-in-Picture) and floating OS-KB
![grafik](https://user-images.githubusercontent.com/45426380/100217632-f8346700-2f13-11eb-81d6-73995702e6bb.png)
Android 10 benefits include in the app/pictures and transparency management. Here is _one shot_ with **Termux** (mc+KB above) and **X11** (check+touch below). The _widget table_ is partially visible at the bottom left - _looks like windows but isn't._

If `ls env ps` is selected, the action:`ls -l;env;ps -eo pid,ppid,comm,etime;pstree;exit 1` _should_ run in termux _**but isn't.**_
Even if several widget actions (9) are selected, _nothing happens._ If **PiP** is _not_ used or minimized for Termux, the action will be carried out _correctly._ I use `shortcut.next-session=ctrl+n` and described this as a _popup_  to `ENTER`. To start the _missing_ actions in Termux _**one `ctrl-n` is necessary**_ so that all selections are displayed in the `DRAWER` and processed step by step.

**Comment by RalfWerner** [comment]
Thanks @xeffyr! The googled PiP documentation does **not** contain any reference to general valid PiP administration (for almost all Apps), but only for a small list of Google apps with a focus on youtube. My four (Q/R) emulations do _**not allow** PiP for Termux & Co._ The management of the PiP list takes place on my Galaxy device in the task list. In addition to the three pictures from my [Doc](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-733700709), I've added camera and gallery here.
![grafik](https://user-images.githubusercontent.com/45426380/102327570-39a1ba80-3f86-11eb-8643-b9b28d151c28.png)
This allows many useful possibilities to operate with termux and the other apps and to follow the _logs-on-device_ in parallel. In **shot** are six apps/tasks and five of which are in the PiP list. _**Can you do that on your device too?**_
This is what the 5 apps look like when they are positioned on the display (startscreen).
![grafik](https://user-images.githubusercontent.com/45426380/102332772-d49d9300-3f8c-11eb-9fa8-ceaa3d56c130.png)

**Comment by RalfWerner** [comment]
### My current idea is to keep the _apt/pkg_ updates in `target 29+` versions
 to create new `bootstrap$a.zip` (test without problem) from the _backups_ after `pkg in/up`, which will be used for building`1)` the next `*.apk` generation but with ELF-Split`3)` and complete _symlink table_ to `nl=../usr/nl`. That would also be possible at _**28**_ and would only require _one_ [symlink](https://github.com/termux/termux-app/issues/1617#issuecomment-695179562) for _**29+**_, `files.so` would then be unnecessary, with an _ARCH-split_, the APK size could be reduced for several ARCHs and with `aapt` in `bootstrap$a.zip` the Version/targetSdk (e.g. 103/29) could be displayed at Termux startup`2)`_**- what do you think of that?**_

`1)` Is _build APK on-device_ with _non-rooted_ devices possible and which packages besides the above are required in addition?
Should `proot` fail in the future, would this be an alternative to _make-loops **have you checked something like that?**_

`2)` If `cat` in `ln -s ../usr u; head -n 4 u/bin/login` _replaced_ by `sed -e s%\!%" $v"%`, the version (`$v` ) can be displayed at termux start. E.g. `v="103, target=29, min=24, device=28/arm7l"` is a possible value here. The last part is simple: `v=...;v+=", device=$(getprop ro.build.version.sdk)/$(uname -m)"`. `...` part requires `pm` and/or path (`$nl`) to the installed Termux `$apk.` Unfortunately, they are _inconsistent_ (known/unknown) on different devices but _related._ If `nl=../../lib` _exists_ it's easy.
If `$nl` is known, is: `apk=$nl/../../base.apk` and revers is: `nl=$(dirname $apk)/lib/*`. _I've no idea, if **both are missing -You?**_

`3)` I got the first _**29+** Version_ from _you @xeffyr_ in Sept. 2019 (with ELF split - flexible and easy to understand).
### check with termux version 103 and 104 (29.12.2020)
For now, I'm just looking at `B=bootstrap-i686.zip`. The two versions should be identical after `pkg up` and `pkg in aapt proot`. To check (30.12) I use `find u/|sort>1;pkg list-in>1i` before and after (with 2/3 instead of 1) the two `pkg` actions.
With `wc -l ?` I've got **103:** 4974->5019->5096 and at **104:** 4993->5018->5095 lines.
With `diff` actions I found differences between the` termux-tool` version and `xxhash` (explainable) and created `A=104.tgz` backup of `../usr` (contains symlink `u/nl`) to get a next generation `$B`. Here with bash script and SYMLINKS/ELF-split.
```
$tar tvf 104.tgz|grep usr/nl
lrwxrwxrwx u0_a138/u0_a138       0 2020-12-29 14:50 usr/nl -> /data/data/com.termux/lib
$ tar xf 104.tgz; S=`find usr -type l`;  ls -l $S>S; wc -l S; rm $S
cd usr; [ ! -e nl ]&& mkdir nl; typeset -i i=0 j=1000; echo "ELF check in `pwd` ..."
while read -r f; do i+=1; if grep -q ELF < <(dd if=$f bs=1 count=3 skip=1 2>/dev/null)
   then j+=1; echo "mv $j.so usr/$f;# $f -> nl/$j.so">>../S; mv $f nl/$j.so; fi
done < <(find bin lib* -type f); j-=1000; echo "$j of $i in $2"; cd ..; exit
```
Then 3731+279 files and 1363 lines in S remain (have to be syntactically converted to SYMLINKS.txt (three of the 1084 symlinks contain `$PREFIX` by mistake).

**Comment by RalfWerner** [comment]
### Backups and _distros_.
`pkg in proot-distro; proot-distro install debian-buster` install _debian_. The process corresponds to `u uu` for _ubuntu_ but without apt/backup function and is saved in `u/ ` instead of `../`, but has the advantage that it is maintained by @termux.
If _fileApp_ shows termux, you can also use `[~] ln -s ../usr/var/lib/proot-distro/installed-rootfs distro`.
This shows the `debian-buster` and `$HOME` is `root` (`/home` is empty and _not_ termux)!
![grafik](https://user-images.githubusercontent.com/45426380/106897349-188e5800-66f3-11eb-824e-12dd17b4aba0.png) 
Here is an example with debian _**and my**_ ubuntu distro in `../uu` with `proot -b $HOME:/home ... ` and with 
`[~] ln -s ../uu/root uh` also _uh_ visible in the _fileApp_ but _**not**_ the symlinks to `/home` (_th_) which is a `bind` (only allows by `ubuntu`) and `*/u/X*` are the Xserver of the _different distros._
![grafik](https://user-images.githubusercontent.com/45426380/106897536-568b7c00-66f3-11eb-9be6-0304bd374be9.png)
With `[/home] ln -s /data/data/com.termux/files t` in _debian_ all _termux/ubuntu_ data known here too.

**Comment by RalfWerner** [comment]
### Backups and _distros_.
`pkg in proot-distro; proot-distro install debian-buster` install _debian_. The process corresponds to `u uu` for _ubuntu_ but without apt/backup function and is saved in `u/ ` instead of `../`, but has the advantage that it is maintained by @termux.
If _fileApp_ shows termux, you can also use `[~] ln -s ../usr/var/lib/proot-distro/installed-rootfs distro`.
This shows the `debian-buster` and `$HOME` is `root` (`/home` is empty and _not_ termux)!
![grafik](https://user-images.githubusercontent.com/45426380/106897349-188e5800-66f3-11eb-824e-12dd17b4aba0.png) 
Here is an example with debian _**and my**_ ubuntu distro in `../uu` with `proot -b $HOME:/home ... ` and with 
`[~] ln -s ../uu/root uh` also _uh_ visible in the _fileApp_ but _**not**_ the symlinks to `/home` (_th_) which is a `bind` (only allows by `ubuntu`) and `*/u/X*` are the Xserver of the _different distros._
![grafik](https://user-images.githubusercontent.com/45426380/106897536-568b7c00-66f3-11eb-9be6-0304bd374be9.png)
With `[/home] ln -s /data/data/com.termux/files t` in _debian_ all _termux/ubuntu_ data known here too.

**Comment by RalfWerner** [comment]
### check of _termux_ with `MANAGE_EXTERNAL_STORAGE` and targetSDK=30 on _29_ and _30_ emulators
![grafik](https://user-images.githubusercontent.com/45426380/108187591-ea574200-710e-11eb-95d0-6f921fcd8b76.png)
above 6 shots (with _S-Pen_ modifications). Because of _build in windows_ (`*w`) the [workaround](https://gitter.im/termux/dev?at=5f5a32d85fde567ffa3ae3ae) is required, as well as depending on the
OS:`git clone https://github.com/termux/termux-app` has difference size to _linux_ (`*l`) in  15770 lines (`diff -r *l *w|grep -c \>`) and in 92 of 121 files (`diff -rq *w *l|wc -l` <cr>). Here is a revised version (copy/paste) for the Termux start.
```bash
P=./:$PATH; PATH=/system/bin:$P;echo 'PS1="\w:$ "; e=echo; f="\`find u/"; w="|wc -l\`";s=storage; sd=/$s/emulated/0
alias  ll="ls -al" f="$e f/l $f -type f$w/$f -type l$w" s="$e u/nl $f$w/$f/nl$w" d="du -hd 1" pp="ps -eo pid,ppid,comm,etime"
t=com.termux; a=Android/data/$t/files; l=/data/data/$t/lib;if [ ! -h u ];then ln -s $sd sd; ln -s ../usr u ;fi
if [ ! -f ~/u/bin/bash ];then cp $l/libfiles.so f; dos2unix f;$e wait ...
   while read -r i; do f=u/${i#*\?};d=`dirname $f`;[ ! -d $d ]&& mkdir -p $d; ln -s $l/${i%\?*} $f; done<f
 fi; f=`df|grep stor|grep \-|cut -d% -f2`; [ -s uu.sh ]&& . uu.sh . .;[ -d storage ]&& return
c=$t.app.reload_style; u/bin/am broadcast --es $c $s -a $c $t; ln -s sd/$a sda; ln -s $f fd; ln -s fd/$a fda'>.bashrc
. .bashrc; PATH=$P
```
to create/start `.bashrc`. I assume _understanding_ of the _bash_ script. The alias **f and s** are used to _check_ `u/` (termux).
In addition to the two [boot-loop adjustments](https://gitter.im/termux/dev?at=601ea58624cd6b60d821b7fc), the [Installer](https://github.com/termux/termux-app/blob/master/app/src/main/java/com/termux/app/TermuxInstaller.javahttps://github.com/termux/termux-packages/blob/master/packages/termux-tools/pkg) an abbreviation to: `..." bin / ") || zipEntryName.startsWith ("lib")) {` necessary for _ELF-Split_ - slightly different for the _Android 10_ branch, which is still _everything_ linked to `u/`.
_Boot loop_ overwites [pkg (apt)](https://github.com/termux/termux-packages/blob/master/packages/termux-tools/pkg#L159), which has some problems and has to restore:
- The smallest thing is that all symlinks in `u/*` point to _readonly_ data. So if all _non-ELF_ are copied (`while`  loop above), it should be solved. But I prefer a _v107 commit_ with only `u/libnl` in `$nl` - in preparation.
- The _remove case_ `un * | rem * | rm | del *)` is no longer displayed in the _Usage_ (meaningless if you know)! The attempt `pkg rm proot-distro` did not work as expected.
### Idea of ​​the _IP-Obj_ on Android:
The sdcard (_download_ also takes place here in _sd_) contains a path e.g. _ip_. This contains _also_ `termux.apk`. The app can be installed and started with a double click and then has all the data and functions for editing and saving the object on the device.

So far, easy! But it can be _complex_ if _this_ also applies to object data and functions. An example from my previous professional work would be the certification of an aircraft. The 32G to 256G on _sd_ are probably not sufficient for this, but for a cargo door modification they are and roughly the same functions are required for both.

An _**IP-Obj**_ that termux developers can understand better deals with checking _**29+**_ functions. In my opinion, this includes _X11_ and _distros_ as well as package updates, backups, _widget/API_ management and data transparency. Idea:
- a `boot*.zip` that fits the `$arch` of the device (i686), is supplemented by at least `proot` and for which a [first ELF-split](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-752725805) took place in `u/libnl` (here 279 files), is created.
- _build_ the `termux.apk` uses _this_ `boot*.zip` file depending on _**28>**targetSDK**<29**_ (here **30**) to create the classic `libbootstrap.so` or to remove `libnl` and save it in `$nl=../../lib` to move.
- Update of all functions described in `i/bashrc` (contains `pkg in ...`). This script is linked with `~/.bashrc` and termux-specific initializations are made from the script: with `. uu.sh . .`
- from this _new_ `u/` (possibly check beforehand with `proot`) a new `boot*.zip` is generated and the _loop_ is repeated with _build_ (above).
- functions that are not included in the termux packages are stored in _distros_ (here _nedit_ in ubuntu and debian).

Up to this point, _these_ **two scripts** are needed, both of which can be saved in `i/` (`ln -s sd/ip i`) and linked in `~/` (i.e. independent of the app), which is also part of `i/` is.
